Make lwaftr bench multiprocess

kbara opened the pull request:
Note that this disables the spawn_csv_stats timer, as discussed last week.

Commit:
Make lwaftr bench multiprocess

Note that this disables the spawn_csv_stats timer, as discussed last week.

wingo commented:
Excellent!

As discussed, if you would like to nail down all the details before landing, here's a list of things to do.

- [ ] Make sure "snabb ps" shows just one instance.

- [ ] Add --cpu arg to "snabb lwaftr bench", like "snabb lwaftr run", and make it propagate that argument to the worker, where the worker will bind to that CPU.  Or perhaps the parent should bind to that CPU as well and then rebind to the first CPU on that NUMA node before forking the workers?  Or even after forking the workers, and perhaps that would allow the CPU binding to propagate to the worker naturally?  That way the parent can verify PCI affinities (see the `numa.check_affinity_for_pci_addresses` calls in `run.lua`).

- [ ] Your change also enables multiple processes with `snabb lwaftr run`.  Probably a good idea to see if it works :)

- [ ] Add a --reconfigurable argument to "snabb lwaftr bench", like "snabb lwaftr run" perhaps?  Eventually we will default to on but this would be an easier way to test both cases.

- [ ] Test `snabb config get` on the parent.

Eventually we need to do some perf testing but I think all the above is enough for the next day or two and certainly enough for landing.

Commit:
Add support for --cpu to bench

The original process binds to the numa node associated with the specified CPU
(if any), and the worker process binds to the specified CPU itself.

kbara commented:
I've added support for --cpu to bench. The main process binds to the right numa node (but not a particular CPU on it), and sets an environment variable with the CPU if specified, which the worker process binds to.

I didn't want to have the main process bind to the specific CPU, because ideally only the data plane will run on that CPU - not the leader or supervisor too.

Commit:
Added bench support for --reconfigurable

Commit:
Added front-end lwaftr tests

Commit:
Make lwaftr run --cpu work with multiprocessing

Commit:
Fixed selftest paths to run from make test

kbara commented:
snabb lwaftr run appears to still work (it runs for the specified duration and exits cleanly). The link report is bogusly empty when run with --reconfigurable, since the link report is being run on the leader. I've also changed --cpu so that it should work with --reconfigurable in snabb lwaftr run.

kbara commented:
```snabb ps``` shows two instances, one for the leader and one for the follower (it does not show the supervisors). Is removing the leader from the snabb ps list really the right thing to do?

Commit:
Added an automated test for snabb config get

wingo commented:
@kbara good question!  we need something to list instances that are running leaders I guess, for the purposes of figuring out instances on which to run "snabb config".  "snabb ps" should probably show whether a process is a worker or a main process.

wingo commented on src/program/lwaftr/selftest.sh:
```diff
@@ -17,8 +17,13 @@ echo "Testing snabb lwaftr bench"
+./snabb lwaftr bench --reconfigurable -D 1 ${TDIR}/icmp_on_fail.conf \
+    ${TDIR}/tcp-frominet-bound.pcap ${TDIR}/tcp-fromb4-ipv6.pcap &
+
+LEADER_PID=$!
```
I think there is a race condition here; you might need to wait until the socket exists, limiting the wait to some number of seconds smaller than the -D argument to snabb lwaftr.

kbara replied:
Plausibly. I originally had a sleep there, then removed it, because it wasn't appearing to be a problem in practice (and sleep is always technically the wrong solution.)

Commit:
Do not show an empty link report in run

wingo reviewed (approved):
Let's do this!
